cmd/labs/project/proxy.go

```diff
@@ -0,0 +1,144 @@
+package project
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/fs"
+	"path/filepath"
+	"strings"
+
+	"github.com/databricks/cli/libs/cmdio"
+	"github.com/databricks/cli/libs/log"
+	"github.com/databricks/cli/libs/process"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+type proxy struct {
+	Entrypoint    `yaml:",inline"`
+	Name          string `yaml:"name"`
+	Description   string `yaml:"description"`
+	TableTemplate string `yaml:"table_template,omitempty"`
+	Flags         []flag `yaml:"flags,omitempty"`
+}
+
+func (cp *proxy) register(parent *cobra.Command) {
+	cmd := &cobra.Command{
+		Use:   cp.Name,
+		Short: cp.Description,
+		RunE:  cp.runE,
+	}
+	parent.AddCommand(cmd)
+	flags := cmd.Flags()
+	for _, flag := range cp.Flags {
+		flag.register(flags)
+	}
+}
+
+func (cp *proxy) runE(cmd *cobra.Command, _ []string) error {
+	err := cp.checkUpdates(cmd)
+	if err != nil {
+		return err
+	}
+	args, err := cp.commandInput(cmd)
+	if err != nil {
+		return err
+	}
+	envs, err := cp.Prepare(cmd)
+	if err != nil {
+		return fmt.Errorf("entrypoint: %w", err)
+	}
+	ctx := cmd.Context()
+	log.Debugf(ctx, "Forwarding subprocess: %s", strings.Join(args, " "))
+	if cp.TableTemplate != "" {
+		return cp.renderJsonAsTable(cmd, args, envs)
+	}
+	err = process.Forwarded(ctx, args,
+		cmd.InOrStdin(),
+		cmd.OutOrStdout(),
+		cmd.ErrOrStderr(),
+		process.WithEnvs(envs))
+	if errors.Is(err, fs.ErrNotExist) && cp.IsPythonProject(ctx) {
+		msg := "cannot find Python %s. Please re-run: databricks labs install %s"
+		return fmt.Errorf(msg, cp.MinPython, cp.Name)
+	}
+	return err
+}
+
+func (cp *proxy) renderJsonAsTable(cmd *cobra.Command, args []string, envs map[string]string) error {
+	var buf bytes.Buffer
+	ctx := cmd.Context()
+	err := process.Forwarded(ctx, args,
+		cmd.InOrStdin(),
+		&buf,
+		cmd.ErrOrStderr(),
+		process.WithEnvs(envs))
+	if err != nil {
+		return err
+	}
+	var anyVal any
+	err = json.Unmarshal(buf.Bytes(), &anyVal)
+	if err != nil {
+		return err
+	}
+	// IntelliJ eagerly replaces tabs with spaces, even though we're not asking for it
+	fixedTemplate := strings.ReplaceAll(cp.TableTemplate, "\\t", "\t")
+	return cmdio.RenderWithTemplate(ctx, anyVal, fixedTemplate)
```

Comment:
Downstream processes must render this themselves. This is too much coupling and leaks out the helpers that we define in cmdio and effectively turns them into a public API. This will break at some point in time.

If you want to do this anyway, you can include a rendered in this package that exports the minimal set of helpers (or none at all) and have a clearly defined contract.

Reply:
https://github.com/databricks/cli/issues/994 added it as follow-up work. There will be iterations on interfaces , anyway - like bool, required, and enum flags - https://github.com/databrickslabs/remorph/blob/main/src/databricks/labs/remorph/cli.py#L18-L29